{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To Do:\n",
    "\n",
    "[ ] Assume mixed infection, take top 10 hits per segment, and try to generate consensus\n",
    "\n",
    "- FluB Yamagata : http://www.ncbi.nlm.nih.gov/nuccore/937172197\n",
    "- FluB Victoria: http://www.ncbi.nlm.nih.gov/nucleotide/937171812\n",
    "\n",
    "- -----\n",
    "\n",
    "[x] Curate flu specific database with 2014-2015 (us only?) to query reads to obtain reference.\n",
    "\n",
    "- Do this on a per segment basis\n",
    "- One thing to note is that there were a lot of reads assigned to the Egypt strain and not florida - maybe because of the flanking regions being present?\n",
    "    - Make the database cds only\n",
    "\n",
    "     [ ] Blast H3N2-70 and H3N2-90 separately\n",
    "     [ ] Use latest run to resolve FluB mutations\n",
    "\n",
    "[ ] Blast MinION reads for the clinical samples and use the hits for a reference\n",
    "\n",
    "- See if the mixed infection data can distinguish the different subtypes\n",
    "    - blastn didn’t do so hot. Only 1009 reads were identified as ssRNA viruses and the subtypes were all over the place.\n",
    "        - blastn -num_threads 40 -db /data1/share/db/nt/nt -query all-6-4_8-13_11-09.pass.fasta -out all-6-4_8-13_11-09.pass.blastn.xml -outfmt 5 -max_target_seqs 1 -culling_limit 2 -evalue 0.00005\n",
    "- Use the references to make a consensus instead of the MiSeq consensus to mimic the practical application of the MinION\n",
    "- Examine run by run to see if the h3n2 data (and lab strains) are interfering with the blast results.\n",
    "\n",
    "[ ] Top variants for each platform and look at the overlap between them\n",
    "\n",
    "[ ] Align to Flu Plasmids\n",
    "\n",
    "- Definitely in the 11-09 run\n",
    "\n",
    "[x] IGV snapshot of quads study of the new data while keeping old data. Look at H1N1\n",
    "\n",
    "[x] Coverage plots of each sequencing run with consensus variants and deletions in it\n",
    "\n",
    "[ ] Variant analysis if possible (a big plus if the result is similar to the MiSeq data).\n",
    "     [x] Top SNV comparison between 11-09 data and MiSeq (map MiSeq and MinION data to the MiSeq consensus sequences and plot the % of the consensus nucleotide. Do this with plasmid data\n",
    "     [x] Percent reference base\n",
    "     [ ] Percent snv\n",
    "     [ ] L1 norm (manhattan distance). This is a way to obtain variance between the MiSeq and MinION variance data.\n",
    "          [ ] MiSeq variant sites (from Tim's snplist script) compared to those positions in MinION (using rebasecalled data):\n",
    "     [ ] Look to see if coverage has an affect - overlay depth?\n",
    "               [ ] 4-20/4-21\n",
    "               [x] 6-4/8-13\n",
    "               [ ] 11-09\n",
    "          [ ] All sites from both platforms combined\n",
    "               [ ] 4-20/4-21\n",
    "               [x] 6-4/8-13\n",
    "               [ ] 11-09\n",
    "\n",
    "[ ] Comparison of old sequencing kit (SK005) vs new sequencing kit (SK006) (plot difference of L1Norm)\n",
    "\n",
    "[x] H1N1 4-20/4-21 vs 11-09\n",
    "\n",
    "[x] FluB 6-4/8-13 vs 11-09\n",
    "\n",
    "[x] FluB-PB1 4-20/4-21 vs 11-09\n",
    "\n",
    "[ ] Compare to MiSeq\n",
    "     [ ] MiSeq vs old kit\n",
    "     [ ] MiSeq vs new kit\n",
    "\n",
    "- Why is the H1N1 NS segment so bad?\n",
    "\n",
    "[ ] Comparison of old metrichor results to new metrichor results (not applicable to run 11-09)\n",
    "\n",
    "[ ] Write script to compare sequence content\n",
    "\n",
    "    - Quality score content?\n",
    "\n",
    "[ ] 4-20/4-21\n",
    "\n",
    "[x] 6-4\n",
    "\n",
    "[x] 8-13\n",
    "\n",
    "- Need to compare in timeline?\n",
    "\n",
    "[x] See if H3N2-70 is distinguishable from H3N2-90 when mixed (use rebasecalled data)\n",
    "     [x] Alignment\n",
    "     [x] Table of variants between the two\n",
    "     [ ] See if there’s any carry over with the H3N2 variants in the alignments\n",
    "\n",
    "- The 6-4 and 8-13 data were merged and H3N2-70 and H3N2-90 were resolved at 100% consensus accuracy (minus the variants that were corrected by the 11-09 run)\n",
    "- Data were aligned to all clinical strains (H3N2-70, H3N2-90, FluB, and pH1N1)\n",
    "- Need figure for this\n",
    "\n",
    "    - Maybe see if they cluster together?\n",
    "    - Coverage plot for now?\n",
    "\n",
    "[x] All the 3 lab reference strains and all the 3 clinical samples at 100% correct at consensus level, compared to the MiSeq data (minimal requirement for the paper)\n",
    "\n",
    "[x] rebasecalled all of the reads\n",
    "\n",
    "[x] H1N1 and FluB-PB1 lab standards to be resequenced.\n",
    "\n",
    "[x] FluB-PB1 clinical to be resequenced\n",
    "\n",
    "[x] 11-9 data was 100% for each sequence\n",
    "\n",
    "[x] Compare FluB clinical consensus and lab standard to see if they’re distinct enough to be run on the same flow cell."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
