{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Consensus Analysis\n",
    "\n",
    "The objective for this analysis is to generate MinION consensus sequences that are 100% identical to either:\n",
    "A) Their reference sequence (only applicable for the lab strain and plasmids)\n",
    "B) Their MiSeq counterpart"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Lab Strain Consensus Analysis\n",
    "For the lab strains, reads were mapped to the references provided by Bin for each subtype."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[bwa_index] fail to open file 'ont2d' : No such file or directory\r\n",
      "[bam_sort] Use -T PREFIX / -o FILE to specify temporary and final output files\r\n",
      "Usage: samtools sort [options...] [in.bam]\r\n",
      "Options:\r\n",
      "  -l INT     Set compression level, from 0 (uncompressed) to 9 (best)\r\n",
      "  -m INT     Set maximum memory per thread; suffix K/M/G recognized [768M]\r\n",
      "  -n         Sort by read name\r\n",
      "  -o FILE    Write final output to FILE rather than standard output\r\n",
      "  -T PREFIX  Write temporary files to PREFIX.nnnn.bam\r\n",
      "  -@, --threads INT\r\n",
      "             Set number of sorting and compression threads [1]\r\n",
      "      --input-fmt-option OPT[=VAL]\r\n",
      "               Specify a single input file format option in the form\r\n",
      "               of OPTION or OPTION=VALUE\r\n",
      "  -O, --output-fmt FORMAT[,OPT[=VAL]]...\r\n",
      "               Specify output format (SAM, BAM, CRAM)\r\n",
      "      --output-fmt-option OPT[=VAL]\r\n",
      "               Specify a single output file format option in the form\r\n",
      "               of OPTION or OPTION=VALUE\r\n",
      "      --reference FILE\r\n",
      "               Reference sequence FASTA FILE [null]\r\n",
      "[E::bwa_idx_load_from_disk] fail to locate the index files\r\n",
      "bash: blasr: command not found\r\n",
      "Settings:\r\n",
      "  Output files: \"all-minion-flu-cds-only.fasta.*.bt2\"\r\n",
      "  Line rate: 6 (line is 64 bytes)\r\n",
      "  Lines per side: 1 (side is 64 bytes)\r\n",
      "  Offset rate: 4 (one in 16)\r\n",
      "  FTable chars: 10\r\n",
      "  Strings: unpacked\r\n",
      "  Max bucket size: default\r\n",
      "  Max bucket size, sqrt multiplier: default\r\n",
      "  Max bucket size, len divisor: 4\r\n",
      "  Difference-cover sample period: 1024\r\n",
      "  Endianness: little\r\n",
      "  Actual local endianness: little\r\n",
      "  Sanity checking: disabled\r\n",
      "  Assertions: disabled\r\n",
      "  Random seed: 0\r\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\r\n",
      "Input files DNA, FASTA:\r\n",
      "  all-minion-flu-cds-only.fasta\r\n",
      "Error: could not open all-minion-flu-cds-only.fasta\r\n",
      "Total time for call to driver() for forward index: 00:00:00\r\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\r\n",
      "Command: bowtie2-build --wrapper basic-0 all-minion-flu-cds-only.fasta all-minion-flu-cds-only.fasta \r\n",
      "Bowtie 2 version 2.2.6 by Ben Langmead (langmea@cs.jhu.edu, www.cs.jhu.edu/~langmea)\r\n",
      "Usage: \r\n",
      "  bowtie2 [options]* -x <bt2-idx> {-1 <m1> -2 <m2> | -U <r>} [-S <sam>]\r\n",
      "\r\n",
      "  <bt2-idx>  Index filename prefix (minus trailing .X.bt2).\r\n",
      "             NOTE: Bowtie 1 and Bowtie 2 indexes are not compatible.\r\n",
      "  <m1>       Files with #1 mates, paired with files in <m2>.\r\n",
      "             Could be gzip'ed (extension: .gz) or bzip2'ed (extension: .bz2).\r\n",
      "  <m2>       Files with #2 mates, paired with files in <m1>.\r\n",
      "             Could be gzip'ed (extension: .gz) or bzip2'ed (extension: .bz2).\r\n",
      "  <r>        Files with unpaired reads.\r\n",
      "             Could be gzip'ed (extension: .gz) or bzip2'ed (extension: .bz2).\r\n",
      "  <sam>      File for SAM output (default: stdout)\r\n",
      "\r\n",
      "  <m1>, <m2>, <r> can be comma-separated lists (no whitespace) and can be\r\n",
      "  specified many times.  E.g. '-U file1.fq,file2.fq -U file3.fq'.\r\n",
      "\r\n",
      "Options (defaults in parentheses):\r\n",
      "\r\n",
      " Input:\r\n",
      "  -q                 query input files are FASTQ .fq/.fastq (default)\r\n",
      "  --qseq             query input files are in Illumina's qseq format\r\n",
      "  -f                 query input files are (multi-)FASTA .fa/.mfa\r\n",
      "  -r                 query input files are raw one-sequence-per-line\r\n",
      "  -c                 <m1>, <m2>, <r> are sequences themselves, not files\r\n",
      "  -s/--skip <int>    skip the first <int> reads/pairs in the input (none)\r\n",
      "  -u/--upto <int>    stop after first <int> reads/pairs (no limit)\r\n",
      "  -5/--trim5 <int>   trim <int> bases from 5'/left end of reads (0)\r\n",
      "  -3/--trim3 <int>   trim <int> bases from 3'/right end of reads (0)\r\n",
      "  --phred33          qualities are Phred+33 (default)\r\n",
      "  --phred64          qualities are Phred+64\r\n",
      "  --int-quals        qualities encoded as space-delimited integers\r\n",
      "\r\n",
      " Presets:                 Same as:\r\n",
      "  For --end-to-end:\r\n",
      "   --very-fast            -D 5 -R 1 -N 0 -L 22 -i S,0,2.50\r\n",
      "   --fast                 -D 10 -R 2 -N 0 -L 22 -i S,0,2.50\r\n",
      "   --sensitive            -D 15 -R 2 -N 0 -L 22 -i S,1,1.15 (default)\r\n",
      "   --very-sensitive       -D 20 -R 3 -N 0 -L 20 -i S,1,0.50\r\n",
      "\r\n",
      "  For --local:\r\n",
      "   --very-fast-local      -D 5 -R 1 -N 0 -L 25 -i S,1,2.00\r\n",
      "   --fast-local           -D 10 -R 2 -N 0 -L 22 -i S,1,1.75\r\n",
      "   --sensitive-local      -D 15 -R 2 -N 0 -L 20 -i S,1,0.75 (default)\r\n",
      "   --very-sensitive-local -D 20 -R 3 -N 0 -L 20 -i S,1,0.50\r\n",
      "\r\n",
      " Alignment:\r\n",
      "  -N <int>           max # mismatches in seed alignment; can be 0 or 1 (0)\r\n",
      "  -L <int>           length of seed substrings; must be >3, <32 (22)\r\n",
      "  -i <func>          interval between seed substrings w/r/t read len (S,1,1.15)\r\n",
      "  --n-ceil <func>    func for max # non-A/C/G/Ts permitted in aln (L,0,0.15)\r\n",
      "  --dpad <int>       include <int> extra ref chars on sides of DP table (15)\r\n",
      "  --gbar <int>       disallow gaps within <int> nucs of read extremes (4)\r\n",
      "  --ignore-quals     treat all quality values as 30 on Phred scale (off)\r\n",
      "  --nofw             do not align forward (original) version of read (off)\r\n",
      "  --norc             do not align reverse-complement version of read (off)\r\n",
      "  --no-1mm-upfront   do not allow 1 mismatch alignments before attempting to\r\n",
      "                     scan for the optimal seeded alignments\r\n",
      "  --end-to-end       entire read must align; no clipping (on)\r\n",
      "   OR\r\n",
      "  --local            local alignment; ends might be soft clipped (off)\r\n",
      "\r\n",
      " Scoring:\r\n",
      "  --ma <int>         match bonus (0 for --end-to-end, 2 for --local) \r\n",
      "  --mp <int>         max penalty for mismatch; lower qual = lower penalty (6)\r\n",
      "  --np <int>         penalty for non-A/C/G/Ts in read/ref (1)\r\n",
      "  --rdg <int>,<int>  read gap open, extend penalties (5,3)\r\n",
      "  --rfg <int>,<int>  reference gap open, extend penalties (5,3)\r\n",
      "  --score-min <func> min acceptable alignment score w/r/t read length\r\n",
      "                     (G,20,8 for local, L,-0.6,-0.6 for end-to-end)\r\n",
      "\r\n",
      " Reporting:\r\n",
      "  (default)          look for multiple alignments, report best, with MAPQ\r\n",
      "   OR\r\n",
      "  -k <int>           report up to <int> alns per read; MAPQ not meaningful\r\n",
      "   OR\r\n",
      "  -a/--all           report all alignments; very slow, MAPQ not meaningful\r\n",
      "\r\n",
      " Effort:\r\n",
      "  -D <int>           give up extending after <int> failed extends in a row (15)\r\n",
      "  -R <int>           for reads w/ repetitive seeds, try <int> sets of seeds (2)\r\n",
      "\r\n",
      " Paired-end:\r\n",
      "  -I/--minins <int>  minimum fragment length (0)\r\n",
      "  -X/--maxins <int>  maximum fragment length (500)\r\n",
      "  --fr/--rf/--ff     -1, -2 mates align fw/rev, rev/fw, fw/fw (--fr)\r\n",
      "  --no-mixed         suppress unpaired alignments for paired reads\r\n",
      "  --no-discordant    suppress discordant alignments for paired reads\r\n",
      "  --no-dovetail      not concordant when mates extend past each other\r\n",
      "  --no-contain       not concordant when one mate alignment contains other\r\n",
      "  --no-overlap       not concordant when mates overlap at all\r\n",
      "\r\n",
      " Output:\r\n",
      "  -t/--time          print wall-clock time taken by search phases\r\n",
      "  --un <path>           write unpaired reads that didn't align to <path>\r\n",
      "  --al <path>           write unpaired reads that aligned at least once to <path>\r\n",
      "  --un-conc <path>      write pairs that didn't align concordantly to <path>\r\n",
      "  --al-conc <path>      write pairs that aligned concordantly at least once to <path>\r\n",
      "  (Note: for --un, --al, --un-conc, or --al-conc, add '-gz' to the option name, e.g.\r\n",
      "  --un-gz <path>, to gzip compress output, or add '-bz2' to bzip2 compress output.)\r\n",
      "  --quiet            print nothing to stderr except serious errors\r\n",
      "  --met-file <path>  send metrics to file at <path> (off)\r\n",
      "  --met-stderr       send metrics to stderr (off)\r\n",
      "  --met <int>        report internal counters & metrics every <int> secs (1)\r\n",
      "  --no-unal          supppress SAM records for unaligned reads\r\n",
      "  --no-head          supppress header lines, i.e. lines starting with @\r\n",
      "  --no-sq            supppress @SQ header lines\r\n",
      "  --rg-id <text>     set read group id, reflected in @RG line and RG:Z: opt field\r\n",
      "  --rg <text>        add <text> (\"lab:value\") to @RG line of SAM header.\r\n",
      "                     Note: @RG line only printed when --rg-id is set.\r\n",
      "  --omit-sec-seq     put '*' in SEQ and QUAL fields for secondary alignments.\r\n",
      "\r\n",
      " Performance:\r\n",
      "  -p/--threads <int> number of alignment threads to launch (1)\r\n",
      "  --reorder          force SAM output order to match order of input reads\r\n",
      "  --mm               use memory-mapped I/O for index; many 'bowtie's can share\r\n",
      "\r\n",
      " Other:\r\n",
      "  --qc-filter        filter out reads that are bad according to QSEQ filter\r\n",
      "  --seed <int>       seed for random number generator (0)\r\n",
      "  --non-deterministic seed rand. gen. arbitrarily instead of using read attributes\r\n",
      "  --version          print version information and quit\r\n",
      "  -h/--help          print this usage message\r\n",
      "***\r\n",
      "Error: Must specify at least one read input with -U/-1/-2\r\n",
      "(ERR): bowtie2-align exited with value 1\r\n",
      "Bowtie 2 version 2.2.6 by Ben Langmead (langmea@cs.jhu.edu, www.cs.jhu.edu/~langmea)\r\n",
      "Usage: \r\n",
      "  bowtie2 [options]* -x <bt2-idx> {-1 <m1> -2 <m2> | -U <r>} [-S <sam>]\r\n",
      "\r\n",
      "  <bt2-idx>  Index filename prefix (minus trailing .X.bt2).\r\n",
      "             NOTE: Bowtie 1 and Bowtie 2 indexes are not compatible.\r\n",
      "  <m1>       Files with #1 mates, paired with files in <m2>.\r\n",
      "             Could be gzip'ed (extension: .gz) or bzip2'ed (extension: .bz2).\r\n",
      "  <m2>       Files with #2 mates, paired with files in <m1>.\r\n",
      "             Could be gzip'ed (extension: .gz) or bzip2'ed (extension: .bz2).\r\n",
      "  <r>        Files with unpaired reads.\r\n",
      "             Could be gzip'ed (extension: .gz) or bzip2'ed (extension: .bz2).\r\n",
      "  <sam>      File for SAM output (default: stdout)\r\n",
      "\r\n",
      "  <m1>, <m2>, <r> can be comma-separated lists (no whitespace) and can be\r\n",
      "  specified many times.  E.g. '-U file1.fq,file2.fq -U file3.fq'.\r\n",
      "\r\n",
      "Options (defaults in parentheses):\r\n",
      "\r\n",
      " Input:\r\n",
      "  -q                 query input files are FASTQ .fq/.fastq (default)\r\n",
      "  --qseq             query input files are in Illumina's qseq format\r\n",
      "  -f                 query input files are (multi-)FASTA .fa/.mfa\r\n",
      "  -r                 query input files are raw one-sequence-per-line\r\n",
      "  -c                 <m1>, <m2>, <r> are sequences themselves, not files\r\n",
      "  -s/--skip <int>    skip the first <int> reads/pairs in the input (none)\r\n",
      "  -u/--upto <int>    stop after first <int> reads/pairs (no limit)\r\n",
      "  -5/--trim5 <int>   trim <int> bases from 5'/left end of reads (0)\r\n",
      "  -3/--trim3 <int>   trim <int> bases from 3'/right end of reads (0)\r\n",
      "  --phred33          qualities are Phred+33 (default)\r\n",
      "  --phred64          qualities are Phred+64\r\n",
      "  --int-quals        qualities encoded as space-delimited integers\r\n",
      "\r\n",
      " Presets:                 Same as:\r\n",
      "  For --end-to-end:\r\n",
      "   --very-fast            -D 5 -R 1 -N 0 -L 22 -i S,0,2.50\r\n",
      "   --fast                 -D 10 -R 2 -N 0 -L 22 -i S,0,2.50\r\n",
      "   --sensitive            -D 15 -R 2 -N 0 -L 22 -i S,1,1.15 (default)\r\n",
      "   --very-sensitive       -D 20 -R 3 -N 0 -L 20 -i S,1,0.50\r\n",
      "\r\n",
      "  For --local:\r\n",
      "   --very-fast-local      -D 5 -R 1 -N 0 -L 25 -i S,1,2.00\r\n",
      "   --fast-local           -D 10 -R 2 -N 0 -L 22 -i S,1,1.75\r\n",
      "   --sensitive-local      -D 15 -R 2 -N 0 -L 20 -i S,1,0.75 (default)\r\n",
      "   --very-sensitive-local -D 20 -R 3 -N 0 -L 20 -i S,1,0.50\r\n",
      "\r\n",
      " Alignment:\r\n",
      "  -N <int>           max # mismatches in seed alignment; can be 0 or 1 (0)\r\n",
      "  -L <int>           length of seed substrings; must be >3, <32 (22)\r\n",
      "  -i <func>          interval between seed substrings w/r/t read len (S,1,1.15)\r\n",
      "  --n-ceil <func>    func for max # non-A/C/G/Ts permitted in aln (L,0,0.15)\r\n",
      "  --dpad <int>       include <int> extra ref chars on sides of DP table (15)\r\n",
      "  --gbar <int>       disallow gaps within <int> nucs of read extremes (4)\r\n",
      "  --ignore-quals     treat all quality values as 30 on Phred scale (off)\r\n",
      "  --nofw             do not align forward (original) version of read (off)\r\n",
      "  --norc             do not align reverse-complement version of read (off)\r\n",
      "  --no-1mm-upfront   do not allow 1 mismatch alignments before attempting to\r\n",
      "                     scan for the optimal seeded alignments\r\n",
      "  --end-to-end       entire read must align; no clipping (on)\r\n",
      "   OR\r\n",
      "  --local            local alignment; ends might be soft clipped (off)\r\n",
      "\r\n",
      " Scoring:\r\n",
      "  --ma <int>         match bonus (0 for --end-to-end, 2 for --local) \r\n",
      "  --mp <int>         max penalty for mismatch; lower qual = lower penalty (6)\r\n",
      "  --np <int>         penalty for non-A/C/G/Ts in read/ref (1)\r\n",
      "  --rdg <int>,<int>  read gap open, extend penalties (5,3)\r\n",
      "  --rfg <int>,<int>  reference gap open, extend penalties (5,3)\r\n",
      "  --score-min <func> min acceptable alignment score w/r/t read length\r\n",
      "                     (G,20,8 for local, L,-0.6,-0.6 for end-to-end)\r\n",
      "\r\n",
      " Reporting:\r\n",
      "  (default)          look for multiple alignments, report best, with MAPQ\r\n",
      "   OR\r\n",
      "  -k <int>           report up to <int> alns per read; MAPQ not meaningful\r\n",
      "   OR\r\n",
      "  -a/--all           report all alignments; very slow, MAPQ not meaningful\r\n",
      "\r\n",
      " Effort:\r\n",
      "  -D <int>           give up extending after <int> failed extends in a row (15)\r\n",
      "  -R <int>           for reads w/ repetitive seeds, try <int> sets of seeds (2)\r\n",
      "\r\n",
      " Paired-end:\r\n",
      "  -I/--minins <int>  minimum fragment length (0)\r\n",
      "  -X/--maxins <int>  maximum fragment length (500)\r\n",
      "  --fr/--rf/--ff     -1, -2 mates align fw/rev, rev/fw, fw/fw (--fr)\r\n",
      "  --no-mixed         suppress unpaired alignments for paired reads\r\n",
      "  --no-discordant    suppress discordant alignments for paired reads\r\n",
      "  --no-dovetail      not concordant when mates extend past each other\r\n",
      "  --no-contain       not concordant when one mate alignment contains other\r\n",
      "  --no-overlap       not concordant when mates overlap at all\r\n",
      "\r\n",
      " Output:\r\n",
      "  -t/--time          print wall-clock time taken by search phases\r\n",
      "  --un <path>           write unpaired reads that didn't align to <path>\r\n",
      "  --al <path>           write unpaired reads that aligned at least once to <path>\r\n",
      "  --un-conc <path>      write pairs that didn't align concordantly to <path>\r\n",
      "  --al-conc <path>      write pairs that aligned concordantly at least once to <path>\r\n",
      "  (Note: for --un, --al, --un-conc, or --al-conc, add '-gz' to the option name, e.g.\r\n",
      "  --un-gz <path>, to gzip compress output, or add '-bz2' to bzip2 compress output.)\r\n",
      "  --quiet            print nothing to stderr except serious errors\r\n",
      "  --met-file <path>  send metrics to file at <path> (off)\r\n",
      "  --met-stderr       send metrics to stderr (off)\r\n",
      "  --met <int>        report internal counters & metrics every <int> secs (1)\r\n",
      "  --no-unal          supppress SAM records for unaligned reads\r\n",
      "  --no-head          supppress header lines, i.e. lines starting with @\r\n",
      "  --no-sq            supppress @SQ header lines\r\n",
      "  --rg-id <text>     set read group id, reflected in @RG line and RG:Z: opt field\r\n",
      "  --rg <text>        add <text> (\"lab:value\") to @RG line of SAM header.\r\n",
      "                     Note: @RG line only printed when --rg-id is set.\r\n",
      "  --omit-sec-seq     put '*' in SEQ and QUAL fields for secondary alignments.\r\n",
      "\r\n",
      " Performance:\r\n",
      "  -p/--threads <int> number of alignment threads to launch (1)\r\n",
      "  --reorder          force SAM output order to match order of input reads\r\n",
      "  --mm               use memory-mapped I/O for index; many 'bowtie's can share\r\n",
      "\r\n",
      " Other:\r\n",
      "  --qc-filter        filter out reads that are bad according to QSEQ filter\r\n",
      "  --seed <int>       seed for random number generator (0)\r\n",
      "  --non-deterministic seed rand. gen. arbitrarily instead of using read attributes\r\n",
      "  --version          print version information and quit\r\n",
      "  -h/--help          print this usage message\r\n",
      "***\r\n",
      "Error: Must specify at least one read input with -U/-1/-2\r\n",
      "(ERR): bowtie2-align exited with value 1\r\n",
      "bash: To: command not found\r\n"
     ]
    }
   ],
   "source": [
    "# MinION data\n",
    "bwa index ont2d all-minion-flu-cds-only.fasta\n",
    "bwa mem -t 40 -x ont2d all-minion-flu-cds-only.fasta flu_4_rebasecalled.fastq |\\\n",
    "samtools view -bS - | \\\n",
    "samtools sort - flu_4_rebasecalled.all-flu-standard-minion && \\\n",
    "samtools index flu_4_rebasecalled.all-flu-standard-minion.bam\n",
    "\n",
    "# PacBio Data\n",
    "blasr Elodie_sidebar_reads_of_insert.fastq all-minion-flu-cds-only.fasta \\\n",
    "-sam -clipping soft -out pacbio-quad-cds-only.bam -nproc 40 -bestn 1\n",
    "\n",
    "# MiSeq Data\n",
    "bowtie2-build all-minion-flu-cds-only.fasta all-minion-flu-cds-only.fasta\n",
    "bowtie2 -p 40 --very-sensitive-local -x all-minion-flu-cds-only.fasta\n",
    "\n",
    "# Ion Data\n",
    "bowtie2 -p 40 --very-sensitive-local -x all-minion-flu-cds-only.fasta\n",
    "\n",
    "# To generate consensus sequences:\n",
    "for i in $(samtools idxstats flu_4_rebasecalled.all-flu-standard-minion.bam | cut -f1 | grep -v '*')\n",
    "do\n",
    "    for x in $(ls *bam | perl -pe 's/\\.bam//g')\n",
    "    do\n",
    "        python bamfile_consensus_generate.py $x.bam $i >> $x-$i.consensus.fasta\n",
    "    done\n",
    "done\n",
    "\n",
    "# To compare the consensus sequences:\n",
    "for i in $(samtools idxstats flu_4_rebasecalled.all-flu-standard-minion.bam | cut -f1 | grep -v '*')\n",
    "do\n",
    "\n",
    "    grep -A1 -h --no-group-separator $i miseq-quad-cds-only-$i.consensus.fasta \\\n",
    "    flu_4_rebasecalled.all-flu-standard-minion-$i.consensus.fasta > miseq-minion-$i.consensus.fasta\\\n",
    "    && python pairwise-pid.py miseq-minion-$i.consensus.fasta fasta > miseq-minion-$i.consensus.variants.txt\n",
    "\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All but FluB-PB1 were resolved with this workflow as there was a suspected sequence-specific error in the PB1 segment. This was corrected with the latest 11-09 run."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clinical Strain Consensus Analysis\n",
    "\n",
    "### First attempt\n",
    "The first run of the consensus analysis was done in the same fashion with the same reference as the lab strains: map the raw reads to the consensus sequences to try and obtain a consensus sequence that is 100% identical to its reference. However, it didn't turn out so good. The main problem was that the workflow was contingent upon a known reference which limits the purpose of using this for diagnostics. It also did not match up with the MiSeq data which suggests something went awry."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Second attempt\n",
    "The second attempt was actually emulate a simple clinical analysis that would be used for diagnostics while using the MiSeq for validation.\n",
    "\n",
    "The first step is to blast the MinION reads to a Flu specific database from [FluDB](http://www.fludb.org/brc/influenza_sequence_search_segment_display.spg?method=ShowCleanSearch&decorator=influenza). This is used to improve runtime by only querying Flu and, further, the subtypes we know are present. The original implementation of this algorithm was limiting the possible strains to the years 2014-2015 since that's the timeline our samples were obtained.\n",
    "\n",
    "<img src=\"flu-db-search-criteria.png\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "makeblastdb -dbtype nucl -in all-flu-2014-2015.fasta\n",
    "\n",
    "# also did this for the 8-13 and 11-9 runs\n",
    "blastn \\\n",
    "-db all-flu-2014-2015.fasta \\\n",
    "-query flu-6-4-rebasecalled.pass.fasta \\\n",
    "-outfmt 5 \\\n",
    "-out flu-6-4-rebasecalled.pass.all-flu-2014-15.blastn.xml \\\n",
    "-evalue 0.00005\\\n",
    "-num_threads 40 \\\n",
    "-culling_limit 2 \\\n",
    "-max_target_seqs 1\n",
    "\n",
    "python read-fludb-blastxml.py flu-6-4-rebasecalled.pass.all-flu-2014-15.blastn.xml > flu-6-4-rebasecalled.pass.all-flu-2014-15.blast.seg-counts.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2-11\n",
    "\n",
    "[ ] Align 11-09 data to MiSeq FluB-PB2 consensus and see if the discrepancy at position 1545 is resolved."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "MinION Flu Analysis\n",
    "\n",
    "For Elodies Presentation:\n",
    "\n",
    "- Compare metrichor version output\n",
    "    - compare sequencing kit output\n",
    "- can use MinION for identification of mixed infections\n",
    "    - subtypes are easily separated\n",
    "    - FluB-PB1 clinical and standard were identifiable after being run through the flowcell\n",
    "    - Can easily distinguish between subtypes\n",
    "    - Maybe distinguish between H3N2-70 and H3N2-90?\n",
    "- SNV calling + haplotypes\n",
    "    - L1 Norm of only MiSeq variants compared to corresponding MinION positions\n",
    "    - L1 Norm of all MiSeq position compared to all MinION positions\n",
    "- Error correction for haplotyping\n",
    "    - I already have a working version of the page rank algorithm to find linked variants\n",
    "\n",
    "To Do:\n",
    "\n",
    "[ ] Assume mixed infection, take top 10 hits per segment, and try to generate consensus\n",
    "\n",
    "- FluB Yamagata : http://www.ncbi.nlm.nih.gov/nuccore/937172197\n",
    "- FluB Victoria: http://www.ncbi.nlm.nih.gov/nucleotide/937171812\n",
    "\n",
    "- -----\n",
    "\n",
    "[x] Curate flu specific database with 2014-2015 (us only?) to query reads to obtain reference.\n",
    "\n",
    "- Do this on a per segment basis\n",
    "- One thing to note is that there were a lot of reads assigned to the Egypt strain and not florida - maybe because of the flanking regions being present?\n",
    "    - Make the database cds only\n",
    "\n",
    "     [ ] Blast H3N2-70 and H3N2-90 separately\n",
    "     [ ] Use latest run to resolve FluB mutations\n",
    "\n",
    "[ ] Blast MinION reads for the clinical samples and use the hits for a reference\n",
    "\n",
    "- See if the mixed infection data can distinguish the different subtypes\n",
    "    - blastn didn’t do so hot. Only 1009 reads were identified as ssRNA viruses and the subtypes were all over the place.\n",
    "        - blastn -num_threads 40 -db /data1/share/db/nt/nt -query all-6-4_8-13_11-09.pass.fasta -out all-6-4_8-13_11-09.pass.blastn.xml -outfmt 5 -max_target_seqs 1 -culling_limit 2 -evalue 0.00005\n",
    "- Use the references to make a consensus instead of the MiSeq consensus to mimic the practical application of the MinION\n",
    "- Examine run by run to see if the h3n2 data (and lab strains) are interfering with the blast results.\n",
    "\n",
    "[ ] Top variants for each platform and look at the overlap between them\n",
    "\n",
    "[ ] Align to Flu Plasmids\n",
    "\n",
    "- Definitely in the 11-09 run\n",
    "\n",
    "[x] IGV snapshot of quads study of the new data while keeping old data. Look at H1N1\n",
    "\n",
    "[x] Coverage plots of each sequencing run with consensus variants and deletions in it\n",
    "\n",
    "[ ] Variant analysis if possible (a big plus if the result is similar to the MiSeq data).\n",
    "     [x] Top SNV comparison between 11-09 data and MiSeq (map MiSeq and MinION data to the MiSeq consensus sequences and plot the % of the consensus nucleotide. Do this with plasmid data\n",
    "     [x] Percent reference base\n",
    "     [ ] Percent snv\n",
    "     [ ] L1 norm (manhattan distance). This is a way to obtain variance between the MiSeq and MinION variance data.\n",
    "          [ ] MiSeq variant sites (from Tim's snplist script) compared to those positions in MinION (using rebasecalled data):\n",
    "     [ ] Look to see if coverage has an affect - overlay depth?\n",
    "               [ ] 4-20/4-21\n",
    "               [x] 6-4/8-13\n",
    "               [ ] 11-09\n",
    "          [ ] All sites from both platforms combined\n",
    "               [ ] 4-20/4-21\n",
    "               [x] 6-4/8-13\n",
    "               [ ] 11-09\n",
    "\n",
    "[ ] Comparison of old sequencing kit (SK005) vs new sequencing kit (SK006) (plot difference of L1Norm)\n",
    "\n",
    "[x] H1N1 4-20/4-21 vs 11-09\n",
    "\n",
    "[x] FluB 6-4/8-13 vs 11-09\n",
    "\n",
    "[x] FluB-PB1 4-20/4-21 vs 11-09\n",
    "\n",
    "[ ] Compare to MiSeq\n",
    "     [ ] MiSeq vs old kit\n",
    "     [ ] MiSeq vs new kit\n",
    "\n",
    "- Why is the H1N1 NS segment so bad?\n",
    "\n",
    "[ ] Comparison of old metrichor results to new metrichor results (not applicable to run 11-09)\n",
    "\n",
    "[ ] Write script to compare sequence content\n",
    "\n",
    "    - Quality score content?\n",
    "\n",
    "[ ] 4-20/4-21\n",
    "\n",
    "[x] 6-4\n",
    "\n",
    "[x] 8-13\n",
    "\n",
    "- Need to compare in timeline?\n",
    "\n",
    "[x] See if H3N2-70 is distinguishable from H3N2-90 when mixed (use rebasecalled data)\n",
    "     [x] Alignment\n",
    "     [x] Table of variants between the two\n",
    "     [ ] See if there’s any carry over with the H3N2 variants in the alignments\n",
    "\n",
    "- The 6-4 and 8-13 data were merged and H3N2-70 and H3N2-90 were resolved at 100% consensus accuracy (minus the variants that were corrected by the 11-09 run)\n",
    "- Data were aligned to all clinical strains (H3N2-70, H3N2-90, FluB, and pH1N1)\n",
    "- Need figure for this\n",
    "\n",
    "    - Maybe see if they cluster together?\n",
    "    - Coverage plot for now?\n",
    "\n",
    "[x] All the 3 lab reference strains and all the 3 clinical samples at 100% correct at consensus level, compared to the MiSeq data (minimal requirement for the paper)\n",
    "\n",
    "[x] rebasecalled all of the reads\n",
    "\n",
    "[x] H1N1 and FluB-PB1 lab standards to be resequenced.\n",
    "\n",
    "[x] FluB-PB1 clinical to be resequenced\n",
    "\n",
    "[x] 11-9 data was 100% for each sequence\n",
    "\n",
    "[x] Compare FluB clinical consensus and lab standard to see if they’re distinct enough to be run on the same flow cell.\n",
    "\n",
    "Interesting Things:\n",
    "\n",
    "- Plot quality scores in FluB-PB1 mutation to see if that’s something\n",
    "    - suspected SSE was found in both the lab standards and the clinical strains!!\n",
    "        - This was seen in both the original and rebasecalled data but not on the new sequencing kit (SK006?)\n",
    "- The newer sequencing kit has a lot more noise than the others. Possibly due to the molecules moving faster through the pores in the new kit (a tradeoff between speed/yield and accuracy) and/or the metrichor base calling algorithm having more information with the old kits\n",
    "\n",
    "Data:\n",
    "\n",
    "Sequencing Runs:\n",
    "\n",
    "- Quads study\n",
    "\n",
    "    - Lab standard H1N1, H3N2, and FluB were sequenced on MiSeq, PacBio, and Ion Torrent. They were pooled individual as well as a mix of all three strains.\n",
    "- MinION 4-20 and 4-21\n",
    "\n",
    "    - Strains sequenced:\n",
    "        - Lab standard H1N1\n",
    "        - Lab standard H3N2\n",
    "        - Lab standard FluB\n",
    "- MinION 6-4\n",
    "\n",
    "    - Strains sequenced:\n",
    "\n",
    "        - clinical pH1N1 (a1)\n",
    "        - clinical FluB 14 (f3)\n",
    "        - clinical H3N2-70 (c1)\n",
    "- MinION 8-13\n",
    "\n",
    "    - Strains sequenced:\n",
    "\n",
    "        - clinical pH1N1 (a1)\n",
    "            - spike-in purified, non-size selected sample to compensate for loss of M/NS\n",
    "        - clinical FluB 14 (f3) NP segment only\n",
    "            - This was selected for because the rest of segments were well covered\n",
    "        - clinical H3N2-90 (f1)\n",
    "            - Different strain because??\n",
    "        - FluB-HA (plasmid)\n",
    "            - This was to see the error rate in the sequencer\n",
    "    - These are all clinical samples. They were resequenced to obtain more data for consensus construction and variant calling.\n",
    "- MinION 11-9\n",
    "\n",
    "    - Strains sequenced\n",
    "\n",
    "        - Lab standard H1N1\n",
    "        - Lab standard FluB-PB1 (plasmid)\n",
    "            - Done to resolve putative SSE\n",
    "        - Clinical FluB 14\n",
    "    - These were resequenced to resolve the errors in the consensus sequences of the previous runs.\n",
    "\n",
    "MiSeq Sample Sheet:\n",
    "\n",
    "\n",
    "\n",
    "L1 Norm comparisons:\n",
    "MiSeq variants only\n",
    "![alt text](miseq-minion-per-pos-manhattan-dists.pdf)\n",
    "all positions\n",
    "\n",
    "old metrichor vs new metrichor manhattan distance of clinical data\n",
    "\n",
    "sk005 vs sk006 manhattan distance\n",
    "positions H1N1-NS:586-599 were removed due to the deletion in the 4-20/21 sequence\n",
    "\n",
    "Flu 11-9 Consensus\n",
    "\n",
    "\n",
    "\n",
    "Mixed Infection Analysis\n",
    "variants are present because this is only the 6-4 8-13 data. These variants have been resolved in the 11-9 data\n",
    "\n",
    "Strain\n",
    "\n",
    "Variants\n",
    "\n",
    "Deletions\n",
    "\n",
    "FluB-HA\n",
    "\n",
    "1\n",
    "\n",
    "10\n",
    "\n",
    "FluB-MP\n",
    "\n",
    "0\n",
    "\n",
    "13\n",
    "\n",
    "FluB-NA\n",
    "\n",
    "0\n",
    "\n",
    "6\n",
    "\n",
    "FluB-NP\n",
    "\n",
    "0\n",
    "\n",
    "7\n",
    "\n",
    "FluB-NS\n",
    "\n",
    "0\n",
    "\n",
    "1\n",
    "\n",
    "FluB-PA\n",
    "\n",
    "0\n",
    "\n",
    "11\n",
    "\n",
    "FluB-PB1\n",
    "\n",
    "1\n",
    "\n",
    "19\n",
    "\n",
    "FluB-PB2\n",
    "\n",
    "1\n",
    "\n",
    "24\n",
    "\n",
    "H1N1-HA\n",
    "\n",
    "0\n",
    "\n",
    "15\n",
    "\n",
    "H1N1-MP\n",
    "\n",
    "0\n",
    "\n",
    "2\n",
    "\n",
    "H1N1-NA\n",
    "\n",
    "0\n",
    "\n",
    "2\n",
    "\n",
    "H1N1-NP\n",
    "\n",
    "0\n",
    "\n",
    "2\n",
    "\n",
    "H1N1-NS\n",
    "\n",
    "0\n",
    "\n",
    "3\n",
    "\n",
    "H1N1-PA\n",
    "\n",
    "0\n",
    "\n",
    "4\n",
    "\n",
    "H1N1-PB1\n",
    "\n",
    "0\n",
    "\n",
    "9\n",
    "\n",
    "H1N1-PB2\n",
    "\n",
    "0\n",
    "\n",
    "6\n",
    "\n",
    "H3N2-70-HA\n",
    "\n",
    "0\n",
    "\n",
    "8\n",
    "\n",
    "H3N2-70-MP\n",
    "\n",
    "0\n",
    "\n",
    "2\n",
    "\n",
    "H3N2-70-NA\n",
    "\n",
    "0\n",
    "\n",
    "8\n",
    "\n",
    "H3N2-70-NP\n",
    "\n",
    "0\n",
    "\n",
    "8\n",
    "\n",
    "H3N2-70-NS\n",
    "\n",
    "0\n",
    "\n",
    "7\n",
    "\n",
    "H3N2-70-PA\n",
    "\n",
    "0\n",
    "\n",
    "7\n",
    "\n",
    "H3N2-70-PB1\n",
    "\n",
    "0\n",
    "\n",
    "17\n",
    "\n",
    "H3N2-70-PB2\n",
    "\n",
    "0\n",
    "\n",
    "13\n",
    "\n",
    "H3N2-90-HA\n",
    "\n",
    "0\n",
    "\n",
    "7\n",
    "\n",
    "H3N2-90-MP\n",
    "\n",
    "0\n",
    "\n",
    "2\n",
    "\n",
    "H3N2-90-NA\n",
    "\n",
    "0\n",
    "\n",
    "7\n",
    "\n",
    "H3N2-90-NP\n",
    "\n",
    "0\n",
    "\n",
    "7\n",
    "\n",
    "H3N2-90-NS\n",
    "\n",
    "0\n",
    "\n",
    "6\n",
    "\n",
    "H3N2-90-PA\n",
    "\n",
    "0\n",
    "\n",
    "11\n",
    "\n",
    "H3N2-90-PB1\n",
    "\n",
    "0\n",
    "\n",
    "17\n",
    "\n",
    "H3N2-90-PB2\n",
    "\n",
    "0\n",
    "\n",
    "9\n",
    "\n",
    "mixed infection coverage plot\n",
    "\n",
    "lab standard coverage plot\n",
    "\n",
    "11-9 coverage plot\n",
    "\n",
    "Percent Consensus base\n",
    "\n",
    "Variants between H3N2-70 and 90\n",
    "\n",
    "Segment\n",
    "\n",
    "Variations\n",
    "\n",
    "HA\n",
    "\n",
    "10\n",
    "\n",
    "NA\n",
    "\n",
    "10\n",
    "\n",
    "NP\n",
    "\n",
    "1\n",
    "\n",
    "NS\n",
    "\n",
    "2\n",
    "\n",
    "MP\n",
    "\n",
    "2\n",
    "\n",
    "PA\n",
    "\n",
    "2\n",
    "\n",
    "PB1\n",
    "\n",
    "6\n",
    "\n",
    "PB2\n",
    "\n",
    "7\n",
    "\n",
    "Blast MinION data to get reference:\n",
    "\n",
    "- megablast performed horribly. Others have reported success with blastn. That was better and the yields are as follows:\n",
    "    - Influenza A virus (A/Egypt/42/2014(H1N1))\n",
    "    - Influenza A virus (A/Singapore/H2013.778/2013(H3N2))\n",
    "    - Influenza B virus (B/Nagasaki/13N070/2014)\n",
    "- These are not what we were expecting so in order to be more precise a manually curated database was constructed\n",
    "- Curated 2014-2015 Flu only database using the following parameters:\n",
    "\n",
    "- Blast workflow was created as follows:\n",
    "    - Blast raw MinION reads to the flu DB\n",
    "    - Extract most frequent segments per subtype\n",
    "        - Maybe change this to the top x% to assess mixed infections/reassortment?\n",
    "\n",
    "        - Used this script:\n",
    "    - Align to selected strains and generate consensus\n",
    "        - Consensus sequences were very good when compared to the MiSeq data\n",
    "\n",
    "        -\n",
    "RefName RefLen  #Variants      #Deletions\n",
    "\n",
    "\n",
    "\n",
    "FluB-NA        1401    0      6\n",
    "\n",
    "\n",
    "\n",
    "FluB-NP        1683    0      7\n",
    "\n",
    "\n",
    "\n",
    "FluB-PA        2181    0      10\n",
    "\n",
    "\n",
    "\n",
    "FluB-PB1        2259    1      18\n",
    "\n",
    "\n",
    "\n",
    "FluB-PB2        2313    2      25\n",
    "\n",
    "\n",
    "\n",
    "H1N1-HA        1701    0      16\n",
    "\n",
    "\n",
    "\n",
    "H1N1-NA        1410    0      2\n",
    "\n",
    "\n",
    "\n",
    "H1N1-NP        1497    0      2\n",
    "\n",
    "\n",
    "\n",
    "H1N1-PA        2151    0      4\n",
    "\n",
    "\n",
    "\n",
    "H1N1-PB1        2274    0      9\n",
    "\n",
    "\n",
    "\n",
    "H1N1-PB2        2280    0      6\n",
    "\n",
    "\n",
    "\n",
    "H3N2-HA        1701    2      7\n",
    "\n",
    "\n",
    "\n",
    "H3N2-MP        759    3      1\n",
    "\n",
    "\n",
    "\n",
    "H3N2-NP        1497    1      7\n",
    "\n",
    "\n",
    "\n",
    "H3N2-PB1        2274    4      16\n",
    "\n",
    "\n",
    "H3N2-PB2        2280    8       10\n",
    "\n",
    "        - Some segments aren’t listed because they were of different size - may be due to different coding sequences selected from the blast results?\n",
    "    - blast the consensus again to see if there is any refinement\n",
    "\n",
    "Data Analysis:\n",
    "\n",
    "- MiSeq data were aligned to the CDS of the sequences that Bin provided\n",
    "\n",
    "- The MiSeq consensus sequences were generated via this script:\n",
    "\n",
    "- All data were then aligned to the MiSeq references and their subsequent consensus sequences were generated via the same script.\n",
    "- Pac bio were aligned via blasr\n",
    "    - /opt/bioinformatics-software/minion-software/blasr/alignment/bin/blasr Elodie_sidebar_reads_of_insert.fastq all-minion-flu-cds-only.fasta -sam -clipping soft -out pacbio-quad-cds-only.bam -nproc 40 -bestn 1\n",
    "- MiSeq data were aligned using ‘bowtie2 –very-sensitive-local –X 2000 ‘\n",
    "    - Insert size was increased beyond what was expected in order to allow for DIs\n",
    "-  Ion data were aligned using ‘bowtie2 –very-sensitive-local’\n",
    "- MinION data were aligned via bwa\n",
    "    - /opt/bioinformatics-software/bwa-0.7.12/bwa mem -t 30 -x ont2d clinical-minion-refs.fasta pass.fasta\n",
    "    - LAST wasn’t working as well\n",
    "\n",
    "- The consensus sequences of each platform were compared using:\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
